### ISSUE TITLE
Offer a @Source annotation to indicate the Ceylon native declaration source  file name associated to a Java implementation source file

### ISSUE THREAD

**Issue by davidfestal** [ceylon/ceylon-compiler#2126]

**Comment by FroMage** [comment]
Why can't we make the association during typechecking? Presumably we have to typecheck both the `native` declaration (empty) and its backend-specific implementation (to check its signature, which we don't do ATM).

**Comment by davidfestal** [comment]
Currently, the Ceylon source is not typechecked : either it is directly compiled from Java or loaded from the binary archive from the model loader, since the model is loaded from binary

**Comment by FroMage** [comment]
Yes, but that's always been wrong and a hack special to the language module. Ideally we should check both to compare the signature, and indeed any generalisation for users such as what @quintesse is working on ATM on ceylon/ceylon-spec#500 would have to check both.

**Comment by davidfestal** [comment]
@FroMage : I'm OK with that. @quintesse was proposing this `@Source` annotation as a simple solution to meet this purpose. But I'm OK, if you prefer let's postpone this for when the whole backend-specific thing will have been  fully thought.

**Comment by FroMage** [comment]
Right, I think we should.

**Comment by davidfestal** [comment]
@FroMage : what you're saying is true for source. By the way correctly compiling *both* the Ceylon-written native header *and* the Java-written native implementation already works, after some tiny changes I made on Tako's work. You can have a look [here](https://github.com/davidfestal/ceylon-natives-with-java-implementation) to see it in action.
However, the case for which I feel the annotation necessary is the case of binary archives. In this case, we only load the model from binaries, and in the CAR, only the native Java declaration has been generated, and there is no information about where the corresponding Ceylon header source code is.
So in this case, inside the IDE, the only way we could do to map the Java binary native declaration with the Ceylon header present in the source archive, would be to typecheck the whole source archive first, and search the Ceylon header declaration in all the source archive phased units. This would be *very* bad in terms of performance, because now we only typecheck, on-the-fly and when necessary, the required source archive phasedunit.

**Comment by davidfestal** [comment]
well probably it would be simpler to produce another mapping text file in the META-INF directory (similar to `mapping.txt`), that contains a mapping from the Java implementation files to the Ceylon header files.
This would be quite simple to add inside the compiler.

**Comment by FroMage** [comment]
So how does this tie in to the new native stuff, @quintesse ?

**Comment by quintesse** [comment]
I don't think it's directly connected. This native stuff is located in a `.ceylon` file after all. I think this has more to do with native code that's implemented in Java, right?
Also because a native declaration now needs to be fully defined in a single source file we have less problems to deal with when it comes to incremental compilation.

**Comment by FroMage** [comment]
OK, so that's something else then. Does the typechecker allow me to implement a native type/function in Java ATM? Like we do in the language module.

**Comment by quintesse** [comment]
Well David has made that work but when removing the ability to do incremental compilation (and have native declarations spread over several files) I also unintentionally broke it. The idea is to make that work again. Maybe David wants to take a look at it again, otherwise when I get back.

**Comment by FroMage** [comment]
I am NOT touching this one guys. @davidfestal and @quintesse figure out if that's still relevant or not now that the `native` stuff has settled down.

**Comment by FroMage** [comment]
Fine, moving to 1.3

**Comment by quintesse** [comment]
It's not something I'm really familiar with, in the end it's something very specific for the IDE.
And I'm not sure how to combine this with your suggestion "Why can't we make the association during typechecking?", because that goes into ModelLoader business which I'm notoriously bad at (the whole reason why right now headers and their implementations have to be in the same file).

**Comment by davidfestal** [comment]
My idea was to implement this myself according to the following comment : https://github.com/ceylon/ceylon-compiler/issues/2126#issuecomment-104290174
which indeed is a bit different from the original issue description.

**Comment by davidfestal** [comment]
so if you want I can create a new issue for this

**Comment by quintesse** [comment]
That is probably a better solution indeed because it would not depend on the programmer adding (and perhaps forgetting to add) the `@Source` element.
